src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1154,519 @@ struct Wallet
+    /**
+    * brief: writeWatchOnlyWallet -  create a new watch-only wallet file with view keys from current wallet
+    * param: wallet_name - name of the current wallet file
+    * param: password - password for new watch-only wallet
+    * outparam: new_keys_file_name - wallet_name + "-watchonly.keys"
+    * note: sets status error on fail
+    */
+    virtual void writeWatchOnlyWallet(const std::string &wallet_name, const std::string &password, std::string &new_keys_file_name) = 0;
+    /**
+    * brief: balancePerSubaddress - get balance for each subaddress in given account
+    * param: index_major - account index
+    * param: strict -
+    * return: [minor subaddress index: amount, ... ]
+    */
+    virtual std::map<std::uint32_t, std::uint64_t> balancePerSubaddress(std::uint32_t index_major, bool strict) const = 0;
+    /**
+    * brief: unlockedBalancePerSubaddress - get unlocked balance for each subaddresses in given account
+    * param: index_major - account index
+    * param: strict -
+    * return: [minor subaddress index: {amount, {blocks to unlock, time to unlock} }, ... ]
+    */
+    virtual std::map<std::uint32_t, std::pair<std::uint64_t, std::pair<std::uint64_t, std::uint64_t>>> unlockedBalancePerSubaddress(std::uint32_t index_major, bool strict) const = 0;
+    // QUESTION : Can anyone help with this comment? There are other ones I'm not satisfied with, so if you have a better suggestion I'll change them.
+    /**
+    * brief: isTransferUnlocked - check if transfer is unlocked
+    * param: unlock_time -
+    * param: block_height -
+    * return: true if transfer is unlocked, else false
+    */
+    virtual bool isTransferUnlocked(std::uint64_t unlock_time, std::uint64_t block_height) const = 0;
+    /**
+    * brief: updatePoolState -
+    * outparam: process_txs - [ [tx, tx_id, double_spend_seen], ... ]
+    * param: refreshed - (default: false)
+    * param: try_incremental - (default: false)
+    * note: sets status error on fail
+    * This public method is typically called to make sure that the wallet's pool state is up-to-date by
+    * clients like simplewallet and the RPC daemon. Before incremental update this was the same method
+    * that 'refresh' also used, but now it's more complicated because for the time being we support
+    * the "old" and the "new" way of updating the pool and because only the 'getblocks' call supports
+    * incremental update but we don't want any blocks here.
+    *
+    * simplewallet does NOT update the pool info during automatic refresh to avoid disturbing interactive
+    * messages and prompts. When it finally calls this method here "to catch up" so to say we can't use
+    * incremental update anymore, because with that we might miss some txs altogether.
+    */
+    virtual void updatePoolState(std::vector<std::tuple<cryptonote::transaction, std::string, bool>> &process_txs, bool refreshed = false, bool try_incremental = false) = 0;
```

Comment:
I am not sure yet what possibilities of "pool info update" the Wallet API has to offer, or should offer, but I am pretty sure that this method in this form is on too low a level of abstraction, like the following `processPoolState` as well.

Reply:
This is too complex for me right now to fully dive into, maybe someone with more experience in this area can comment?

Reply:
I try to follow the general principle of YAGNI, especially when making APIs. Don't add it unless someone asks for it. A use case for processing abstractions of the pool state might pop up in the future at some point, but it also might not. Until then, these functions are internal business logic and shouldn't be exposed to the API.

Reply:
If we don't add them to the API, how would the [wallet-rpc](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/wallet_rpc_server.cpp#L2855) & [wallet-cli](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/simplewallet/simplewallet.cpp#L8586) call this function, if the goal is to get rid of all the wallet2 function calls? Sorry if I'm missing something.

Reply:
I checked, and it's used indeed, so it's not a case of YAGNI like @jeffro256 suspected, and after I saw that @moneromooo-monero fixed an information leak a few years back when using remote daemons where somehow this method and the next for processing the pool state were involved, I think it's best to just add those to the interface and keep them "public".

Reply:
I think that the function `update_pool_state`, with `process_pool_state` are both leaky abstractions as written. Both times they are used outside of `wallet2`, they are called like so:

```
std::vector<std::tuple<cryptonote::transaction, crypto::hash, bool>> process_txs;
m_wallet->update_pool_state(process_txs);
if (!process_txs.empty())
  m_wallet->process_pool_state(process_txs);
```

This sequence of statements basically means "rescan the pool" and IMO should be abstracted into a method as such. Calling `update_pool_state` without an immediate corresponding call to `process_pool_state` will almost always be a mistake, as this will update the fetching marker without actually scanning the transactions and getting up-to-date information about your balance. So I'd add a method to the `Wallet` interface called something like `refreshPoolOnly` which calls the above statements, and drop both `updatePoolState` and `processPoolState` from the interface.